tensordict/prototype/tensorclass.py

```diff
@@ -532,25 +537,166 @@ def _batch_size_setter(self, new_size: torch.Size) -> None:
     self._tensordict._batch_size_setter(new_size)
 
 
+def _any(self, dim: int = None):
+    """A recursive implementation of `any()` over the tensorclass leaves.
+
+    If the `dim` arg is passed, the resulting tensorclass will have
+    this dimension removed.
+
+    Args:
+        dim (int, optional): if provided, ``any`` will run over that dimension and reduce it.
+            When traversing the tensorclass leaves,  negative dims will be turned to positive
+            ones compared with the root tensorclass' batch-size to avoid any clash
+            of dimensions.
+            Defaults to None, ie full reduction in a single boolean value.
+
+    Examples:
+        >>> @tensorclass
+        ... class MyClass:
+        ...      X: Tensor
+        ...      y: "MyClass"
+        ...
+        >>> c = MyClass(
+        ...     torch.randn(3, 4),
+        ...     MyClass(torch.randn(3, 4, 1), None, batch_size=[3, 4, 1]),
+        ...     batch_size=[3, 4])
+        >>> _ = c.any()  # bool
+        >>> # one can work with negative dimensions, which will be converted to positive dims for the root tensorclass
+        >>> assert (c.any(-1) == c.any(1)).all()
+        >>> assert c.any(-1).y.batch_size == c.any(1).y.batch_size
+
+    """
+    if dim is not None and dim < 0:
+        dim = self.batch_dims + dim
+
+    if dim is not None:
+        non_tensor = {
+            key: None if not is_tensorclass(obj) else obj.any(dim=dim)
+            for key, obj in self._non_tensordict.items()
+        }
+        return self._from_tensordict(
+            self._tensordict.any(dim=dim),
+            non_tensor,
+        )
+    return self._tensordict.any() or any(
+        value.any() for value in self._non_tensordict.values() if is_tensorclass(value)
+    )
+
+
+def _state_dict(self):
```

Comment:
why in this PR? I don't see tests for these

Reply:
because when introducing the non-tensor data feature, we broke `all` which was working out of the box, but now was causing some silent error in the state-dict test
Basically, before all the tensor data was stored in a single tensordict which made it simple to fall back on tensordict methods.
Now we will have to check one-by-one that all methods have a proper fallback since they are not automatically routed towards `self._non_tensordict`